framework/include/base/MooseFunctorArguments.h

```diff
@@ -136,4 +137,41 @@ using ElemQpArg = std::tuple<const libMesh::Elem *, unsigned int, const QBase *>
  * - The quadrature rule that can be used to initialize the functor on the given element and side
  */
 using ElemSideQpArg = std::tuple<const libMesh::Elem *, unsigned int, unsigned int, const QBase *>;
+
+/**
+ * Temporal argument for evaluating functors
+ */
+struct TimeArg
+{
+  /**
+   * Prevent implicit conversions from boolean to avoid users accidentally constructing a time
+   * argument when they meant to construct a skewness argument, etc.
+   */
+  TimeArg(bool) = delete;
+
+  TimeArg(unsigned int state_in) : state(state_in), iteration_type(SolutionIterationType::Time) {}
+
+  TimeArg(unsigned int state_in, SolutionIterationType iteration_type_in)
+    : state(state_in), iteration_type(iteration_type_in)
+  {
+  }
+
+  /// The state. Zero represents the most recent state, so for any kind of iteration type, a zero
+  /// state represents the current state, e.g. current solution
+  unsigned int state;
+
+  /// The solution iteration type, e.g. time or nonlinear
+  SolutionIterationType iteration_type;
+
+private:
+  TimeArg() : state(0), iteration_type(SolutionIterationType::Time) {}
+
+  friend TimeArg currentTimeFunctorArg();
+};
+
+inline TimeArg
+currentTimeFunctorArg()
+{
+  return {};
+}
```

Comment:
should we have:
previousTimeFunctorArg()
and
previousNonlinearFunctorArg() 
predefined